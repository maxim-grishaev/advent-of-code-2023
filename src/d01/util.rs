pub fn take_line_digits(line: &str) -> u32 {
    let mut l: u32 = 10;
    let mut r: u32 = 10;
    // iterate over the line's chars
    for (_i, c) in line.chars().enumerate() {
        let num = match c.to_digit(10) {
            None => continue,
            Some(n) => n,
        };

        if l == 10 {
            l = num;
            r = num;
        } else {
            r = num;
        }
    }
    return (l * 10 + r) as u32;
}
