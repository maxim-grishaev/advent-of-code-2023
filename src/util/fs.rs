use std::path::Path;

pub fn read_file(cur_file: &str, sibling_path: &str) -> String {
    let cur_dir = Path::new(cur_file).parent().unwrap();
    let ref input_path = cur_dir.join(sibling_path);
    // println!("file: {}", input_path.clone().to_str().unwrap());
    return std::fs::read_to_string(input_path)
        .expect("Something went wrong reading the file {cur_file}/{sibling_path}");
}
