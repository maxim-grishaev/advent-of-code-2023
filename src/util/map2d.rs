#![allow(unused)]

use std::fmt::{Debug, Formatter, Result};

#[derive(Debug)]
pub struct Coord2D {
    pub row: usize,
    pub col: usize,
}

impl Coord2D {
    pub fn new(row: usize, col: usize) -> Coord2D {
        Coord2D { row, col }
    }

    pub fn cell_distance(&self, other: &Coord2D) -> usize {
        (self.col as isize - other.col as isize).unsigned_abs()
            + (self.row as isize - other.row as isize).unsigned_abs()
    }
}

#[derive(Debug)]
pub struct Size {
    pub cols_count: usize,
    pub rows_count: usize,
}

#[derive(Debug)]
pub struct Lines2D {
    pub rows: Vec<usize>,
    pub cols: Vec<usize>,
}

pub type Vec2D<T> = Vec<Vec<T>>;

// #[derive(Debug)]
pub struct Map2D<T> {
    pub raw: Vec2D<T>,
}

const NEW_LINE: &str = "\n";
impl Debug for Map2D<char> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        let s = self
            .raw
            .iter()
            .map(|line| line.iter().collect::<String>())
            .collect::<Vec<String>>()
            .join(NEW_LINE);
        write!(f, "Map2D<char>\n{}", s)
    }
}

impl Map2D<char> {
    pub fn from_str(str: &str) -> Map2D<char> {
        Map2D::new(
            str.trim()
                .lines()
                .map(|line| line.chars().collect())
                .collect(),
        )
    }
}

impl<T> Map2D<T> {
    pub fn new(data: Vec<Vec<T>>) -> Map2D<T> {
        Map2D { raw: data }
    }

    pub fn get_val_at(&self, y_row: usize, x_col: usize) -> Option<&T> {
        self.raw.get(y_row).and_then(|line| line.get(x_col))
    }

    pub fn get_val_by_coord(&self, coord: &Coord2D) -> Option<&T> {
        self.get_val_at(coord.row, coord.col)
    }

    pub fn size(&self) -> Size {
        let height = self.raw.len();
        let width = self.raw.get(0).map(|line| line.len()).unwrap_or(0);
        Size {
            rows_count: height,
            cols_count: width,
        }
    }

    pub fn find_coords<F>(&self, is_found: F) -> Vec<Coord2D>
    where
        F: Fn(&T) -> bool,
    {
        let mut coords = vec![];
        self.for_each(|item, coord| {
            if is_found(item) {
                coords.push(coord);
            }
        });
        coords
    }

    pub fn for_each<F>(&self, mut clbk: F)
    where
        F: FnMut(&T, Coord2D),
    {
        for (y, row) in self.raw.iter().enumerate() {
            for (x, item) in row.iter().enumerate() {
                clbk(item, Coord2D { col: x, row: y });
            }
        }
    }

    pub fn find_lines_has_all_of<F>(&self, is_good: F) -> Lines2D
    where
        F: Fn(&T, Coord2D) -> bool,
    {
        let size = self.size();
        find_lines(size.rows_count, size.cols_count, |coord| {
            self.get_val_by_coord(&coord)
                .map_or(false, |c| is_good(c, coord))
        })
    }
}

fn bools_to_indexes(v: Vec<bool>) -> Vec<usize> {
    v.into_iter()
        .enumerate()
        .filter_map(|(idx, b)| match b {
            true => Some(idx),
            false => None,
        })
        .collect()
}

#[allow(clippy::needless_range_loop)]
fn find_lines<F>(rows_count: usize, cols_count: usize, is_found: F) -> Lines2D
where
    F: Fn(Coord2D) -> bool,
{
    let mut rows = vec![true; rows_count];
    let mut cols = vec![true; cols_count];
    for y in 0..rows_count {
        for x in 0..cols_count {
            if !is_found(Coord2D { row: y, col: x }) {
                rows[y] = false;
                cols[x] = false;
            }
        }
    }

    Lines2D {
        rows: bools_to_indexes(rows),
        cols: bools_to_indexes(cols),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn for_each() {
        let m2d = Map2D::new(vec![vec![1, 2], vec![3, 4]]);
        let mut sum = 0;
        m2d.for_each(|item, _coord| {
            sum += item;
        });
        assert_eq!(sum, 10);
    }

    const TEST_MAP_TINY: &str = r#"
..
1.
"#;
    #[test]
    fn find_lines_tiny() {
        let map2d = Map2D::from_str(TEST_MAP_TINY);
        println!("{:?}", map2d);
        let lines = map2d.find_lines_has_all_of(|c, _| c == &'.');
        assert_eq!(lines.rows, vec![0], "rows");
        assert_eq!(lines.cols, vec![1], "cols");
    }

    const TEST_MAP: &str = r#"
.....
1....
..2..
.....
....3
"#;
    #[test]
    fn find_lines() {
        let map2d = Map2D::from_str(TEST_MAP);
        let lines = map2d.find_lines_has_all_of(|c, _| c == &'.');
        assert_eq!(lines.rows, vec![0, 3], "rows");
        assert_eq!(lines.cols, vec![1, 3], "cols");
    }

    #[test]
    fn get_val_at() {
        let m2d = Map2D::new(vec![vec![1, 2], vec![3, 4]]);
        assert_eq!(m2d.get_val_at(0, 0), Some(&1));
        assert_eq!(m2d.get_val_at(0, 1), Some(&2));
        assert_eq!(m2d.get_val_at(1, 0), Some(&3));
        assert_eq!(m2d.get_val_at(1, 1), Some(&4));
        assert_eq!(m2d.get_val_at(2, 0), None);
        assert_eq!(m2d.get_val_at(0, 2), None);
    }
}
