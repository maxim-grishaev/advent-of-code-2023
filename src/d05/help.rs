use crate::util;

pub type ID = u64;
pub type AlmMapCfgLine = (ID, ID, ID);

#[derive(Debug)]
pub struct AlmMap {
    pub title: String,
    pub config: Vec<AlmMapCfgLine>,
}

#[derive(Debug)]
pub struct Almanac {
    pub seeds: Vec<ID>,
    pub maps: Vec<AlmMap>,
}

pub fn read_almanac() -> String {
    util::fs::read_file(file!(), "almanac.txt")
}

pub fn parse_almanac(alm_str: &str) -> Almanac {
    let mut sections = alm_str.split("\n\n");

    let mut seeds_split = sections
        .next()
        .unwrap() // seeds line: "seeds: 79 14 55 13"
        .split(":");
    seeds_split.next(); // seeds label: "seeds"
    let seeds = seeds_split
        .next()
        .unwrap() // seeds string: "79 14 55 13"
        .split_whitespace()
        .map(|s| s.parse::<ID>().unwrap())
        .collect::<Vec<ID>>();

    let mut maps: Vec<AlmMap> = vec![];
    loop {
        let sect = sections.next();
        if sect.is_none() {
            break;
        }

        let mut sect_iter = sect.unwrap().lines();
        let title = sect_iter.next().unwrap(); // title line: "seed-to-soil map:"
        let config_nums = sect_iter.map(|line| {
            let triplet = line
                .split_whitespace()
                .map(|s| s.parse::<ID>().unwrap())
                .collect::<Vec<ID>>();
            if triplet.len() != 3 {
                panic!("triplet.len() != 3: {} {}", title, line);
            }
            (triplet[0], triplet[1], triplet[2])
        });
        maps.push(AlmMap {
            title: title.to_string(),
            config: config_nums.collect(),
        });
    }
    Almanac { seeds, maps }
}

/// The rest of the almanac contains a list of maps which describe how to convert numbers from a source category into numbers in a destination category.
/// That is, the section that starts with seed-to-soil map: describes how to convert a seed number (the source) to a soil number (the destination).
/// This lets the gardener and his team know which soil to use with which seeds, which water to use with which fertilizer, and so on.
///
/// Rather than list every source number and its corresponding destination number one by one, the maps describe entire ranges of numbers that can be converted.
/// Each line within a map contains three numbers: the destination range start, the source range start, and the range length.
///
/// Consider again the example seed-to-soil map:
/// ```
/// 50 98 2
/// 52 50 48
/// ```
/// The first line has a destination range start of 50, a source range start of 98, and a range length of 2.
/// This line means that the source range starts at 98 and contains two values: 98 and 99.
/// The destination range is the same length, but it starts at 50, so its two values are 50 and 51.
/// With this information, you know that seed number 98 corresponds to soil number 50 and that seed number 99 corresponds to soil number 51.
///
/// The second line means that the source range starts at 50 and contains 48 values: 50, 51, ..., 96, 97.
/// This corresponds to a destination range starting at 52 and also containing 48 values: 52, 53, ..., 98, 99.
/// So, seed number 53 corresponds to soil number 55.
///
/// Any source numbers that aren't mapped correspond to the same destination number.
/// So, seed number 10 corresponds to soil number 10.
pub fn get_dest_by_src(src: &ID, amap: &AlmMap) -> ID {
    // let mut dest = *src;
    let dest = *src;
    for (dest_start, src_start, len) in amap.config.iter() {
        if src_start <= &dest && dest < *src_start + *len {
            // dest = dest_start + val - src_start;
            return dest_start + dest - src_start;
        }
    }
    dest
}

pub fn get_destination(seed: &ID, alm: &Almanac) -> ID {
    let mut val = *seed;
    for map in alm.maps.iter() {
        val = get_dest_by_src(&val, map);
    }
    val
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    const AL_STR: &str = r#"
seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
    "#;

    fn get_almanac() -> Almanac {
        parse_almanac(AL_STR.trim())
    }

    #[test]
    fn test_run() {
        assert_debug_snapshot!(get_almanac());
    }

    fn get_destination_path(seed: &ID, alm: &Almanac) -> Vec<ID> {
        let mut val = vec![*seed];
        for map in alm.maps.iter() {
            val.push(get_dest_by_src(val.last().unwrap(), map));
        }
        val
    }

    #[test]
    fn get_destination_test() {
        // Seed 79, soil 81, fertilizer 81, water 81, light 74, temperature 78, humidity 78, location 82.
        assert_eq!(
            get_destination_path(&79, &get_almanac()),
            [79, 81, 81, 81, 74, 78, 78, 82]
        );
        // Seed 14, soil 14, fertilizer 53, water 49, light 42, temperature 42, humidity 43, location 43.
        assert_eq!(
            get_destination_path(&14, &get_almanac()),
            [14, 14, 53, 49, 42, 42, 43, 43]
        );
        // Seed 55, soil 57, fertilizer 57, water 53, light 46, temperature 82, humidity 82, location 86.
        assert_eq!(
            get_destination_path(&55, &get_almanac()),
            [55, 57, 57, 53, 46, 82, 82, 86]
        );
        // Seed 13, soil 13, fertilizer 52, water 41, light 34, temperature 34, humidity 35, location 35.
        assert_eq!(
            get_destination_path(&13, &get_almanac()),
            [13, 13, 52, 41, 34, 34, 35, 35]
        );
    }
}
