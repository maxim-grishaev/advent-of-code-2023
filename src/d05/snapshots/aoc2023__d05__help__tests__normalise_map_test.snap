---
source: src/d05/help.rs
expression: "get_almanac().maps.iter().map(normalise_map).collect::<Vec<_>>()"
---
[
    {
        85: 83,
        59: 57,
        95: 93,
        81: 79,
        78: 76,
        52: 50,
        53: 51,
        86: 84,
        70: 68,
        87: 85,
        97: 95,
        89: 87,
        83: 81,
        54: 52,
        55: 53,
        51: 99,
        65: 63,
        90: 88,
        96: 94,
        77: 75,
        66: 64,
        79: 77,
        91: 89,
        62: 60,
        92: 90,
        94: 92,
        76: 74,
        72: 70,
        56: 54,
        82: 80,
        99: 97,
        93: 91,
        84: 82,
        67: 65,
        73: 71,
        75: 73,
        71: 69,
        68: 66,
        98: 96,
        64: 62,
        88: 86,
        57: 55,
        61: 59,
        74: 72,
        63: 61,
        69: 67,
        50: 98,
        60: 58,
        58: 56,
        80: 78,
    },
    {
        22: 37,
        9: 24,
        41: 2,
        46: 7,
        7: 22,
        48: 9,
        52: 13,
        11: 26,
        10: 25,
        20: 35,
        23: 38,
        17: 32,
        3: 18,
        30: 45,
        44: 5,
        14: 29,
        35: 50,
        16: 31,
        5: 20,
        8: 23,
        34: 49,
        31: 46,
        26: 41,
        38: 53,
        43: 4,
        45: 6,
        50: 11,
        19: 34,
        37: 52,
        49: 10,
        1: 16,
        29: 44,
        4: 19,
        53: 14,
        47: 8,
        25: 40,
        28: 43,
        42: 3,
        0: 15,
        21: 36,
        2: 17,
        18: 33,
        33: 48,
        51: 12,
        12: 27,
        24: 39,
        27: 42,
        32: 47,
        6: 21,
        39: 0,
        13: 28,
        36: 51,
        40: 1,
        15: 30,
    },
    {
        6: 17,
        28: 39,
        39: 50,
        13: 24,
        50: 54,
        21: 32,
        58: 8,
        11: 22,
        17: 28,
        41: 52,
        40: 51,
        47: 5,
        0: 11,
        31: 42,
        43: 1,
        46: 4,
        8: 19,
        7: 18,
        37: 48,
        16: 27,
        52: 56,
        51: 55,
        27: 38,
        54: 58,
        35: 46,
        60: 10,
        38: 49,
        36: 47,
        48: 6,
        12: 23,
        55: 59,
        9: 20,
        32: 43,
        33: 44,
        57: 7,
        42: 0,
        45: 3,
        2: 13,
        30: 41,
        44: 2,
        14: 25,
        10: 21,
        34: 45,
        3: 14,
        15: 26,
        56: 60,
        5: 16,
        24: 35,
        26: 37,
        49: 53,
        23: 34,
        29: 40,
        20: 31,
        4: 15,
        18: 29,
        53: 57,
        25: 36,
        1: 12,
        19: 30,
        22: 33,
        59: 9,
    },
    {
        59: 66,
        86: 93,
        35: 42,
        18: 25,
        64: 71,
        79: 86,
        20: 27,
        90: 20,
        25: 32,
        44: 51,
        67: 74,
        68: 75,
        78: 85,
        34: 41,
        39: 46,
        82: 89,
        71: 78,
        88: 18,
        50: 57,
        56: 63,
        63: 70,
        94: 24,
        23: 30,
        28: 35,
        60: 67,
        92: 22,
        33: 40,
        41: 48,
        29: 36,
        30: 37,
        40: 47,
        19: 26,
        80: 87,
        43: 50,
        48: 55,
        21: 28,
        76: 83,
        84: 91,
        89: 19,
        91: 21,
        93: 23,
        24: 31,
        47: 54,
        65: 72,
        49: 56,
        46: 53,
        72: 79,
        74: 81,
        77: 84,
        69: 76,
        57: 64,
        87: 94,
        75: 82,
        38: 45,
        70: 77,
        26: 33,
        32: 39,
        62: 69,
        51: 58,
        22: 29,
        61: 68,
        85: 92,
        53: 60,
        81: 88,
        52: 59,
        31: 38,
        73: 80,
        66: 73,
        83: 90,
        54: 61,
        36: 43,
        27: 34,
        45: 52,
        37: 44,
        42: 49,
        55: 62,
        58: 65,
    },
    {
        73: 69,
        86: 50,
        74: 70,
        97: 61,
        75: 71,
        55: 87,
        85: 49,
        53: 85,
        62: 94,
        91: 55,
        84: 48,
        57: 89,
        77: 73,
        68: 64,
        50: 82,
        58: 90,
        63: 95,
        89: 53,
        92: 56,
        94: 58,
        70: 66,
        46: 78,
        90: 54,
        71: 67,
        80: 76,
        64: 96,
        82: 46,
        48: 80,
        47: 79,
        60: 92,
        83: 47,
        56: 88,
        61: 93,
        87: 51,
        96: 60,
        98: 62,
        88: 52,
        76: 72,
        78: 74,
        66: 98,
        93: 57,
        95: 59,
        81: 45,
        49: 81,
        65: 97,
        54: 86,
        59: 91,
        72: 68,
        79: 75,
        52: 84,
        45: 77,
        67: 99,
        99: 63,
        69: 65,
        51: 83,
    },
    {
        5: 4,
        9: 8,
        40: 39,
        16: 15,
        1: 0,
        14: 13,
        60: 59,
        13: 12,
        3: 2,
        18: 17,
        55: 54,
        28: 27,
        35: 34,
        31: 30,
        59: 58,
        67: 66,
        66: 65,
        10: 9,
        11: 10,
        34: 33,
        24: 23,
        2: 1,
        23: 22,
        8: 7,
        6: 5,
        43: 42,
        21: 20,
        20: 19,
        46: 45,
        27: 26,
        29: 28,
        30: 29,
        26: 25,
        7: 6,
        15: 14,
        22: 21,
        64: 63,
        4: 3,
        38: 37,
        48: 47,
        56: 55,
        25: 24,
        57: 56,
        62: 61,
        65: 64,
        49: 48,
        50: 49,
        0: 69,
        51: 50,
        52: 51,
        54: 53,
        19: 18,
        37: 36,
        63: 62,
        45: 44,
        69: 68,
        39: 38,
        32: 31,
        42: 41,
        17: 16,
        41: 40,
        47: 46,
        53: 52,
        68: 67,
        44: 43,
        58: 57,
        33: 32,
        12: 11,
        61: 60,
        36: 35,
    },
    {
        57: 94,
        58: 95,
        87: 83,
        92: 88,
        70: 66,
        68: 64,
        66: 62,
        77: 73,
        80: 76,
        72: 68,
        63: 59,
        67: 63,
        60: 56,
        61: 57,
        79: 75,
        78: 74,
        94: 90,
        75: 71,
        84: 80,
        59: 96,
        83: 79,
        85: 81,
        89: 85,
        64: 60,
        82: 78,
        71: 67,
        88: 84,
        90: 86,
        62: 58,
        96: 92,
        69: 65,
        86: 82,
        91: 87,
        76: 72,
        81: 77,
        93: 89,
        95: 91,
        73: 69,
        74: 70,
        65: 61,
        56: 93,
    },
]
